src/crypto/CHIPCryptoPALTinyCrypt.cpp

```diff
@@ -1691,5 +1691,85 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
     return error;
 }
 
+namespace {
+
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+int x509_string_cmp(const mbedtls_x509_buf * a, const mbedtls_x509_buf * b)
+{
+    if (a->CHIP_CRYPTO_PAL_PRIVATE_X509(tag) == b->CHIP_CRYPTO_PAL_PRIVATE_X509(tag) &&
+        a->CHIP_CRYPTO_PAL_PRIVATE_X509(len) == b->CHIP_CRYPTO_PAL_PRIVATE_X509(len) &&
+        memcmp(a->CHIP_CRYPTO_PAL_PRIVATE_X509(p), b->CHIP_CRYPTO_PAL_PRIVATE_X509(p), b->CHIP_CRYPTO_PAL_PRIVATE_X509(len)) == 0)
+    {
+        return 0;
+    }
+    return -1;
+}
+
+CHIP_ERROR ExtractSubjectFromX509Cert(const ByteSpan & certificate, mbedtls_x509_buf & subject)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    int result       = 0;
+    mbedtls_x509_crt mbedCertificate;
+
+    ReturnErrorCodeIf(certificate.empty(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    mbedtls_x509_crt_init(&mbedCertificate);
+    result = mbedtls_x509_crt_parse(&mbedCertificate, Uint8::to_const_uchar(certificate.data()), certificate.size());
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    subject = mbedCertificate.CHIP_CRYPTO_PAL_PRIVATE_X509(subject_raw);
+
+exit:
+    _log_mbedTLS_error(result);
+    mbedtls_x509_crt_free(&mbedCertificate);
+
+    return error;
+}
+#endif // defined(MBEDTLS_X509_CRT_PARSE_C)
+
+} // namespace
+
+CHIP_ERROR ReplaceCertIfResignedCertFound(const ByteSpan & referenceCertificate, const ByteSpan * candidateCertificates,
+                                          size_t candidateCertificatesCount, ByteSpan & outCertificate)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+    mbedtls_x509_buf referenceSubject;
+    mbedtls_x509_buf candidateSubject;
+    uint8_t referenceSKIDBuf[kSubjectKeyIdentifierLength];
+    uint8_t candidateSKIDBuf[kSubjectKeyIdentifierLength];
+    MutableByteSpan referenceSKID(referenceSKIDBuf);
+    MutableByteSpan candidateSKID(candidateSKIDBuf);
+
+    outCertificate = referenceCertificate;
+
+    ReturnErrorCodeIf(candidateCertificates == nullptr || candidateCertificatesCount == 0, CHIP_NO_ERROR);
+
+    ReturnErrorOnFailure(ExtractSubjectFromX509Cert(referenceCertificate, referenceSubject));
+    ReturnErrorOnFailure(ExtractSKIDFromX509Cert(referenceCertificate, referenceSKID));
+
+    for (size_t i = 0; i < candidateCertificatesCount; i++)
+    {
+        const ByteSpan candidateCertificate = candidateCertificates[i];
+
+        ReturnErrorOnFailure(ExtractSubjectFromX509Cert(candidateCertificate, candidateSubject));
+        ReturnErrorOnFailure(ExtractSKIDFromX509Cert(candidateCertificate, candidateSKID));
+
+        if (referenceSKID.data_equal(candidateSKID) && x509_string_cmp(&referenceSubject, &candidateSubject) == 0)
+        {
+            outCertificate = candidateCertificate;
+            return CHIP_NO_ERROR;
+        }
+    }
+
+#else
+    (void) referenceCertificate;
```

Comment:
Should this case really silently violate the API contract for this function?

Reply:
This implementation matches all other method implementations in this file. I would suggest to review and fix it as part of other PR.